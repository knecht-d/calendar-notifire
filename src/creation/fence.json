{
  "tags": ["creation"],
  "exports": ["index"],
  "imports": ["external", "gateways", "useCases", "interfaces"]
}
