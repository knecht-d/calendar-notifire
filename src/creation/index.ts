export * from "./GenericFactory";
export * from "./LocalFactory";
export * from "./Builder";
