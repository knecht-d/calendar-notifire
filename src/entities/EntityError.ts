export class EntityError extends Error {
    constructor(public key: EntityErrorCode) {
        super(`${key}`);

        // Set the prototype explicitly.
        Object.setPrototypeOf(this, EntityError.prototype);
    }
}

export enum EntityErrorCode {
    CHAT_NOT_EXISTING = "CHAT_NOT_EXISTING",
    CHAT_ALREADY_EXISTING = "CHAT_ALREADY_EXISTING",
    MISSING_PRIVILEGES = "MISSING_PRIVILEGES",
    LAST_ADMIN = "LAST_ADMIN",
    NO_ADMIN = "NO_ADMIN",
    TRIGGER_NOT_DEFINED = "TRIGGER_NOT_DEFINED",
}
