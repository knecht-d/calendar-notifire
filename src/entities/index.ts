export * from "./Chat";
export * from "./Chats";
export * from "./RecurrenceRule";
export * from "./TimeFrame";
