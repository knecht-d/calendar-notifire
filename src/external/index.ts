export * from "./chat";
export * from "./storage";
