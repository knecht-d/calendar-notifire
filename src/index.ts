/* istanbul ignore file */
import Telegraf from "telegraf";

const chatId = 42;
const token = "<token>";
const bot = new Telegraf(token);

// Default commands
bot.start(ctx => ctx.reply(`Welcome ${ctx.chat?.first_name} (${ctx.chat?.id})`));
bot.help(ctx => ctx.reply("Send me a sticker"));

// Custom commands
bot.command("test", ({ reply }) => reply("passed"));

// Specific message types
bot.on("sticker", ctx => ctx.reply("👍"));

// Specific texts
bot.hears("hi", ctx => ctx.reply("Hey there"));

// Start
bot.launch().catch(console.error);

// Custom message
bot.telegram.sendMessage(chatId, "I'm Back!").catch(console.error);
