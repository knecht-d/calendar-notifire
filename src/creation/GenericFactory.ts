import { AbstractCalendar, AbstractChat, AbstractStorage, AbstractTimer } from "../external";

export abstract class GenericFactory<CalendarSetup, StorageSetup, ChatSetup> {
    constructor(
        protected classes: {
            calendar: new (setupData: CalendarSetup) => AbstractCalendar<CalendarSetup>;
            storage: new (setupData: StorageSetup) => AbstractStorage<StorageSetup>;
            chat: new (setupData: ChatSetup) => AbstractChat<ChatSetup>;
            timer: new () => AbstractTimer;
        },
    ) {}
    createCalendar(setupData: CalendarSetup): AbstractCalendar<CalendarSetup> {
        return new this.classes.calendar(setupData);
    }
    createChat(setupData: ChatSetup): AbstractChat<ChatSetup> {
        return new this.classes.chat(setupData);
    }
    createStorage(setupData: StorageSetup): AbstractStorage<StorageSetup> {
        return new this.classes.storage(setupData);
    }
    createTimer(): AbstractTimer {
        return new this.classes.timer();
    }
}
