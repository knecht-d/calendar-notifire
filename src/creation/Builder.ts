import {
    CalendarGateway,
    CommunicationController,
    CommunicationPresenter,
    PeristenceGateway,
    TriggerGateway,
} from "../gateways";
import { InitializeChatImpl, ReminderImpl, UpdateConfigImpl, StartAssistantImpl } from "../useCases";
import { GenericFactory } from "./GenericFactory";

export class Builder<CalendarSetup, StorageSetup, ChatSetup> {
    constructor(private factory: GenericFactory<CalendarSetup, StorageSetup, ChatSetup>) {}

    build(setupData: { calendar: CalendarSetup; storage: StorageSetup; chatData: ChatSetup }) {
        // Create External
        const calendar = this.factory.createCalendar(setupData.calendar);
        const chat = this.factory.createChat(setupData.chatData);
        const storage = this.factory.createStorage(setupData.storage);
        const timer = this.factory.createTimer();

        // Create Gateways
        const calendarGW = new CalendarGateway();
        const communicationController = new CommunicationController();
        const communicationPresenter = new CommunicationPresenter();
        const persitenceGW = new PeristenceGateway();
        const triggerGW = new TriggerGateway();

        // Create Use Cases
        const initChat = new InitializeChatImpl(communicationPresenter, persitenceGW);
        const reminder = new ReminderImpl(calendarGW, communicationPresenter);
        const start = new StartAssistantImpl(triggerGW, persitenceGW);
        const updateConfig = new UpdateConfigImpl(communicationPresenter, triggerGW, persitenceGW);

        // Initialize Gateways
        calendarGW.init({ calendarConnector: calendar });
        communicationController.init({
            useCases: { update: updateConfig, init: initChat },
            presenter: communicationPresenter,
        });
        communicationPresenter.init({ communication: chat });
        persitenceGW.init({ persistence: storage });
        triggerGW.init({ triggerConfig: timer, reminder: reminder });

        // Initialize External
        chat.init(communicationController);
        timer.init(triggerGW);

        // Load persisted chats.
        start.execute();

        return {
            calendar,
            chat,
            storage,
            timer,
            start,
        };
    }
}
