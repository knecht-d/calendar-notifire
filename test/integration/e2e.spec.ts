import { Builder } from "../../src/creation";
import { EntityError, EntityErrorCode } from "../../src/entities/EntityError";
import { ILogger, Logger, LogLevels } from "../../src/external";
import { MockCalendar, MockChat, MockFactory, MockStorage, MockTimer } from "../mocks";

let logger: Logger;
let calendar: MockCalendar;
let chat: MockChat;
let storage: MockStorage;
let timer: MockTimer;
jest.mock("../mocks", () => {
    const actual = jest.requireActual("../mocks");
    return {
        ...actual,
        MockFactory: jest.fn().mockImplementation(() => ({
            createCalendar: (logger: ILogger, setupData: any) => {
                calendar = new actual.MockCalendar(logger, setupData);
                return calendar;
            },
            createChat: (logger: ILogger, setupData: any) => {
                chat = new actual.MockChat(logger, setupData);
                return chat;
            },
            createStorage: (logger: ILogger, setupData: any) => {
                storage = new actual.MockStorage(logger, setupData);
                return storage;
            },
            createTimer: (logger: ILogger, setupData: any) => {
                timer = new actual.MockTimer(logger, setupData);
                return timer;
            },
        })),
    };
});
jest.mock("../../src/external/logging");
describe("End to end test", () => {
    beforeAll(async () => {
        const mockFactory = new MockFactory();
        const builder = new Builder(mockFactory);
        await builder.build({
            calendar: { events: [] },
            chatData: {},
            storage: {},
            loggger: { level: LogLevels.error },
        });
        logger = (Logger as jest.Mock<Logger>).mock.instances[0];
    });
    beforeEach(() => {
        jest.clearAllMocks();
        timer.getNextExecution.mockImplementation((id: string) => {
            const defaultValues: { [id: string]: Date } = {
                "chat1|trigger1": new Date(2020, 7, 15, 17, 45),
            };
            return defaultValues[id] || new Date(2020, 0, 1, 12, 30);
        });
    });
    describe("initialization", () => {
        describe("Not initialized", () => {
            describe("admin use cases", () => {
                it("add admin should fail if chat is not initialized", async () => {
                    await chat.fireAddAdmin("chat1", "user1", "user2");
                    expect(chat.send).toHaveBeenCalledTimes(1);
                    expect(chat.send).toHaveBeenCalledWith(
                        "chat1",
                        "Hinzufügen von user2 zu Administratoren fehlgeschlagen.\nChat existiert nicht",
                    );
                    expect(storage.storedData).toEqual({});
                });
                it("remove admin should fail if chat is not initialized", async () => {
                    await chat.fireRemoveAdmin("chat1", "user1", "user2");
                    expect(chat.send).toHaveBeenCalledTimes(1);
                    expect(chat.send).toHaveBeenCalledWith(
                        "chat1",
                        "Entfernen von user2 aus Administratoren fehlgeschlagen.\nChat existiert nicht",
                    );
                    expect(storage.storedData).toEqual({});
                });
            });
            describe("reminder use case", () => {
                it("reminder should fail if chat is not initialized", async () => {
                    await timer.fireTrigger("chat1|trigger1");
                    expect(chat.send).not.toHaveBeenCalled();
                    expect(logger.error).toHaveBeenCalledWith(
                        "ReminderImpl",
                        new EntityError(EntityErrorCode.CHAT_NOT_EXISTING),
                    );
                });
            });
            describe("trigger use cases", () => {
                it("delete trigger should fail if chat is not initialized", async () => {
                    await chat.fireDelete("chat1", "user1", "trigger1");
                    expect(chat.send).toHaveBeenCalledTimes(1);
                    expect(chat.send).toHaveBeenCalledWith(
                        "chat1",
                        "Löschen von trigger1 fehlgeschlagen.\nChat existiert nicht",
                    );
                    expect(storage.storedData).toEqual({});
                });
                it("read trigger should fail if chat is not initialized", async () => {
                    await chat.fireRead("chat1");
                    expect(chat.send).toHaveBeenCalledTimes(1);
                    expect(chat.send).toHaveBeenCalledWith(
                        "chat1",
                        "Lesen der Konfiguration fehlgeschlagen.\nChat existiert nicht",
                    );
                    expect(storage.storedData).toEqual({});
                });
                it("set trigger should fail if chat is not initialized", async () => {
                    await chat.fireSet("chat1", "user1", "trigger1", {
                        recurrence: {
                            type: "m",
                            dayOfMonth: 15,
                            hour: 17,
                            minute: 30,
                        },
                        frame: {
                            begin: {
                                month: {
                                    value: 1,
                                    fixed: false,
                                },
                                day: {
                                    value: 1,
                                    fixed: true,
                                },
                                hour: {
                                    value: 0,
                                    fixed: true,
                                },
                                minute: {
                                    value: 0,
                                    fixed: true,
                                },
                            },
                            end: {
                                month: {
                                    value: 2,
                                    fixed: false,
                                },
                                day: {
                                    value: 1,
                                    fixed: true,
                                },
                                hour: {
                                    value: 0,
                                    fixed: true,
                                },
                                minute: {
                                    value: 0,
                                    fixed: true,
                                },
                            },
                        },
                    });
                    expect(chat.send).toHaveBeenCalledTimes(1);
                    expect(chat.send).toHaveBeenCalledWith(
                        "chat1",
                        "Setzen von trigger1 fehlgeschlagen.\nChat existiert nicht",
                    );
                    expect(storage.storedData).toEqual({});
                });
            });
        });

        it("chat should be initialized", async () => {
            await chat.fireInitChat("chat1", "user1");
            expect(chat.send).toHaveBeenCalledTimes(1);
            expect(chat.send).toHaveBeenCalledWith("chat1", "Initialisierung des Chats erfolgreich.");
            expect(storage.storedData).toEqual({ chat1: '{"administrators":["user1"],"triggerSettings":{}}' });
        });
        it("chat should not be initialized twice", async () => {
            await chat.fireInitChat("chat1", "anotherUser");
            expect(chat.send).toHaveBeenCalledTimes(1);
            expect(chat.send).toHaveBeenCalledWith(
                "chat1",
                "Initialisierung des Chats fehlgeschlagen.\nChat existiert bereits",
            );
            expect(storage.storedData).toEqual({ chat1: '{"administrators":["user1"],"triggerSettings":{}}' });
        });
    });

    describe("admins", () => {
        it("should add a new admin", async () => {
            await chat.fireAddAdmin("chat1", "user1", "user2");
            expect(chat.send).toHaveBeenCalledTimes(1);
            expect(chat.send).toHaveBeenCalledWith("chat1", "user2 erfolgreich zu Administratoren hinzugefügt.");
            expect(storage.storedData).toEqual({
                chat1: '{"administrators":["user1","user2"],"triggerSettings":{}}',
            });
        });
        it("should fail to add a new admin if current user is not an admin", async () => {
            await chat.fireAddAdmin("chat1", "noAdmin", "user3");
            expect(chat.send).toHaveBeenCalledTimes(1);
            expect(chat.send).toHaveBeenCalledWith(
                "chat1",
                "Hinzufügen von user3 zu Administratoren fehlgeschlagen.\nFehlende Berechtigung",
            );
            expect(storage.storedData).toEqual({
                chat1: '{"administrators":["user1","user2"],"triggerSettings":{}}',
            });
        });
        it("should fail to remove an admin if current user is not an admin", async () => {
            await chat.fireRemoveAdmin("chat1", "noAdmin", "user1");
            expect(chat.send).toHaveBeenCalledTimes(1);
            expect(chat.send).toHaveBeenCalledWith(
                "chat1",
                "Entfernen von user1 aus Administratoren fehlgeschlagen.\nFehlende Berechtigung",
            );
            expect(storage.storedData).toEqual({
                chat1: '{"administrators":["user1","user2"],"triggerSettings":{}}',
            });
        });
        it("should fail to remove an admin if to be removed user is not an admin", async () => {
            await chat.fireRemoveAdmin("chat1", "user1", "noAdmin");
            expect(chat.send).toHaveBeenCalledTimes(1);
            expect(chat.send).toHaveBeenCalledWith(
                "chat1",
                "Entfernen von noAdmin aus Administratoren fehlgeschlagen.\nPerson gehört nicht zur Administration",
            );
            expect(storage.storedData).toEqual({
                chat1: '{"administrators":["user1","user2"],"triggerSettings":{}}',
            });
        });
        it("should remove an admin", async () => {
            await chat.fireRemoveAdmin("chat1", "user1", "user2");
            expect(chat.send).toHaveBeenCalledTimes(1);
            expect(chat.send).toHaveBeenCalledWith("chat1", "user2 erfolgreich von Administratoren entfernt.");
            expect(storage.storedData).toEqual({
                chat1: '{"administrators":["user1"],"triggerSettings":{}}',
            });
        });
        it("should fail to remove an admin if it is the last one", async () => {
            await chat.fireRemoveAdmin("chat1", "user1", "user1");
            expect(chat.send).toHaveBeenCalledTimes(1);
            expect(chat.send).toHaveBeenCalledWith(
                "chat1",
                "Entfernen von user1 aus Administratoren fehlgeschlagen.\nLetzer Eintrag kann nicht entfernt werden",
            );
            expect(storage.storedData).toEqual({
                chat1: '{"administrators":["user1"],"triggerSettings":{}}',
            });
        });
    });

    describe("trigger settings", () => {
        it("should set a new trigger", async () => {
            await chat.fireSet("chat1", "user1", "trigger1", {
                recurrence: {
                    type: "m",
                    dayOfMonth: 15,
                    hour: 17,
                    minute: 30,
                },
                frame: {
                    begin: {
                        month: {
                            value: 1,
                            fixed: false,
                        },
                        day: {
                            value: 1,
                            fixed: true,
                        },
                        hour: {
                            value: 0,
                            fixed: true,
                        },
                        minute: {
                            value: 0,
                            fixed: true,
                        },
                    },
                    end: {
                        month: {
                            value: 2,
                            fixed: false,
                        },
                        day: {
                            value: 1,
                            fixed: true,
                        },
                        hour: {
                            value: 0,
                            fixed: true,
                        },
                        minute: {
                            value: 0,
                            fixed: true,
                        },
                    },
                },
            });
            expect(chat.send).toHaveBeenCalledTimes(1);
            expect(chat.send).toHaveBeenCalledWith("chat1", "Setzen von trigger1 erfolgreich.");
            expect(timer.triggers["chat1|trigger1"]).toEqual("30 17 15 * *");
            expect(storage.storedData).toEqual({
                chat1:
                    '{"administrators":["user1"],"triggerSettings":{"trigger1":{"frame":{"begin":{"month":{"value":1,"fixed":false},"day":{"value":1,"fixed":true},"hour":{"value":0,"fixed":true},"minute":{"value":0,"fixed":true}},"end":{"month":{"value":2,"fixed":false},"day":{"value":1,"fixed":true},"hour":{"value":0,"fixed":true},"minute":{"value":0,"fixed":true}}},"recurrence":{"type":"m","day":15,"hour":17,"minute":30}}}}',
            });
        });
        it("should fail to set a new trigger if user is not an admin", async () => {
            await chat.fireSet("chat1", "noAdmin", "trigger2", {
                recurrence: {
                    type: "m",
                    dayOfMonth: 15,
                    hour: 17,
                    minute: 30,
                },
                frame: {
                    begin: {
                        month: {
                            value: 1,
                            fixed: false,
                        },
                        day: {
                            value: 1,
                            fixed: true,
                        },
                        hour: {
                            value: 0,
                            fixed: true,
                        },
                        minute: {
                            value: 0,
                            fixed: true,
                        },
                    },
                    end: {
                        month: {
                            value: 2,
                            fixed: false,
                        },
                        day: {
                            value: 1,
                            fixed: true,
                        },
                        hour: {
                            value: 0,
                            fixed: true,
                        },
                        minute: {
                            value: 0,
                            fixed: true,
                        },
                    },
                },
            });
            expect(chat.send).toHaveBeenCalledTimes(1);
            expect(chat.send).toHaveBeenCalledWith(
                "chat1",
                "Setzen von trigger2 fehlgeschlagen.\nFehlende Berechtigung",
            );
            expect(timer.triggers["chat1|trigger2"]).toBeUndefined();
            expect(storage.storedData).toEqual({
                chat1:
                    '{"administrators":["user1"],"triggerSettings":{"trigger1":{"frame":{"begin":{"month":{"value":1,"fixed":false},"day":{"value":1,"fixed":true},"hour":{"value":0,"fixed":true},"minute":{"value":0,"fixed":true}},"end":{"month":{"value":2,"fixed":false},"day":{"value":1,"fixed":true},"hour":{"value":0,"fixed":true},"minute":{"value":0,"fixed":true}}},"recurrence":{"type":"m","day":15,"hour":17,"minute":30}}}}',
            });
        });
        it("should update an existing trigger", async () => {
            await chat.fireSet("chat1", "user1", "trigger1", {
                recurrence: {
                    type: "m",
                    dayOfMonth: 15,
                    hour: 17,
                    minute: 45,
                },
                frame: {
                    begin: {
                        month: {
                            value: 1,
                            fixed: false,
                        },
                        day: {
                            value: 1,
                            fixed: true,
                        },
                        hour: {
                            value: 0,
                            fixed: true,
                        },
                        minute: {
                            value: 0,
                            fixed: true,
                        },
                    },
                    end: {
                        month: {
                            value: 2,
                            fixed: false,
                        },
                        day: {
                            value: 1,
                            fixed: true,
                        },
                        hour: {
                            value: 0,
                            fixed: true,
                        },
                        minute: {
                            value: 0,
                            fixed: true,
                        },
                    },
                },
            });
            expect(chat.send).toHaveBeenCalledTimes(1);
            expect(chat.send).toHaveBeenCalledWith("chat1", "Setzen von trigger1 erfolgreich.");
            expect(timer.triggers["chat1|trigger1"]).toEqual("45 17 15 * *");
            expect(storage.storedData).toEqual({
                chat1:
                    '{"administrators":["user1"],"triggerSettings":{"trigger1":{"frame":{"begin":{"month":{"value":1,"fixed":false},"day":{"value":1,"fixed":true},"hour":{"value":0,"fixed":true},"minute":{"value":0,"fixed":true}},"end":{"month":{"value":2,"fixed":false},"day":{"value":1,"fixed":true},"hour":{"value":0,"fixed":true},"minute":{"value":0,"fixed":true}}},"recurrence":{"type":"m","day":15,"hour":17,"minute":45}}}}',
            });
        });
        it("should add a additional trigger", async () => {
            await chat.fireSet("chat1", "user1", "trigger2", {
                recurrence: {
                    type: "m",
                    dayOfMonth: 15,
                    hour: 17,
                    minute: 30,
                },
                frame: {
                    begin: {
                        month: {
                            value: 1,
                            fixed: false,
                        },
                        day: {
                            value: 1,
                            fixed: true,
                        },
                        hour: {
                            value: 0,
                            fixed: true,
                        },
                        minute: {
                            value: 0,
                            fixed: true,
                        },
                    },
                    end: {
                        month: {
                            value: 2,
                            fixed: false,
                        },
                        day: {
                            value: 1,
                            fixed: true,
                        },
                        hour: {
                            value: 0,
                            fixed: true,
                        },
                        minute: {
                            value: 0,
                            fixed: true,
                        },
                    },
                },
            });
            expect(chat.send).toHaveBeenCalledTimes(1);
            expect(chat.send).toHaveBeenCalledWith("chat1", "Setzen von trigger2 erfolgreich.");
            expect(timer.triggers["chat1|trigger1"]).toEqual("45 17 15 * *");
            expect(timer.triggers["chat1|trigger2"]).toEqual("30 17 15 * *");
            expect(storage.storedData).toEqual({
                chat1:
                    '{"administrators":["user1"],"triggerSettings":{"trigger1":{"frame":{"begin":{"month":{"value":1,"fixed":false},"day":{"value":1,"fixed":true},"hour":{"value":0,"fixed":true},"minute":{"value":0,"fixed":true}},"end":{"month":{"value":2,"fixed":false},"day":{"value":1,"fixed":true},"hour":{"value":0,"fixed":true},"minute":{"value":0,"fixed":true}}},"recurrence":{"type":"m","day":15,"hour":17,"minute":45}},"trigger2":{"frame":{"begin":{"month":{"value":1,"fixed":false},"day":{"value":1,"fixed":true},"hour":{"value":0,"fixed":true},"minute":{"value":0,"fixed":true}},"end":{"month":{"value":2,"fixed":false},"day":{"value":1,"fixed":true},"hour":{"value":0,"fixed":true},"minute":{"value":0,"fixed":true}}},"recurrence":{"type":"m","day":15,"hour":17,"minute":30}}}}',
            });
        });
        it("should delete an existing trigger", async () => {
            await chat.fireDelete("chat1", "user1", "trigger2");
            expect(chat.send).toHaveBeenCalledTimes(1);
            expect(chat.send).toHaveBeenCalledWith("chat1", "Löschen von trigger2 erfolgreich.");
            expect(timer.triggers["chat1|trigger1"]).toEqual("45 17 15 * *");
            expect(timer.triggers["chat1|trigger2"]).toBeUndefined();
            expect(storage.storedData).toEqual({
                chat1:
                    '{"administrators":["user1"],"triggerSettings":{"trigger1":{"frame":{"begin":{"month":{"value":1,"fixed":false},"day":{"value":1,"fixed":true},"hour":{"value":0,"fixed":true},"minute":{"value":0,"fixed":true}},"end":{"month":{"value":2,"fixed":false},"day":{"value":1,"fixed":true},"hour":{"value":0,"fixed":true},"minute":{"value":0,"fixed":true}}},"recurrence":{"type":"m","day":15,"hour":17,"minute":45}}}}',
            });
        });
        it("should fail to delete an existing trigger if user is not an admin", async () => {
            await chat.fireDelete("chat1", "nonAdmin", "trigger1");
            expect(chat.send).toHaveBeenCalledTimes(1);
            expect(chat.send).toHaveBeenCalledWith(
                "chat1",
                "Löschen von trigger1 fehlgeschlagen.\nFehlende Berechtigung",
            );
            expect(timer.triggers["chat1|trigger1"]).toEqual("45 17 15 * *");
            expect(storage.storedData).toEqual({
                chat1:
                    '{"administrators":["user1"],"triggerSettings":{"trigger1":{"frame":{"begin":{"month":{"value":1,"fixed":false},"day":{"value":1,"fixed":true},"hour":{"value":0,"fixed":true},"minute":{"value":0,"fixed":true}},"end":{"month":{"value":2,"fixed":false},"day":{"value":1,"fixed":true},"hour":{"value":0,"fixed":true},"minute":{"value":0,"fixed":true}}},"recurrence":{"type":"m","day":15,"hour":17,"minute":45}}}}',
            });
        });
        it("should read the config even if user is not an admin", async () => {
            await chat.fireRead("chat1");
            expect(chat.send).toHaveBeenCalledTimes(1);
            expect(chat.send).toMatchSnapshot();
        });
    });

    describe("reminder", () => {
        it("should send the events if a trigger fires", async () => {
            jest.spyOn(calendar, "getEvents").mockResolvedValueOnce([
                {
                    start: new Date(2020, 7, 15, 18, 0),
                    end: new Date(2020, 7, 15, 19, 0),
                    title: "Test",
                    description: "Dies ist ein Test",
                    location: "Überall",
                },
            ]);
            await timer.fireTrigger("chat1|trigger1");
            expect(chat.send).toHaveBeenCalledTimes(1);
            expect(chat.send).toHaveBeenCalledWith(
                "chat1",
                "Termine:\nTest (Überall):\n    15.8.2020 18:00 - 19:00\n    Dies ist ein Test",
            );
            expect(storage.storedData).toEqual({
                chat1:
                    '{"administrators":["user1"],"triggerSettings":{"trigger1":{"frame":{"begin":{"month":{"value":1,"fixed":false},"day":{"value":1,"fixed":true},"hour":{"value":0,"fixed":true},"minute":{"value":0,"fixed":true}},"end":{"month":{"value":2,"fixed":false},"day":{"value":1,"fixed":true},"hour":{"value":0,"fixed":true},"minute":{"value":0,"fixed":true}}},"recurrence":{"type":"m","day":15,"hour":17,"minute":45}}}}',
            });
        });
        it("should not send the events if a trigger fires but there are no events", async () => {
            await timer.fireTrigger("chat1|trigger1");
            expect(chat.send).not.toHaveBeenCalled();
            expect(storage.storedData).toEqual({
                chat1:
                    '{"administrators":["user1"],"triggerSettings":{"trigger1":{"frame":{"begin":{"month":{"value":1,"fixed":false},"day":{"value":1,"fixed":true},"hour":{"value":0,"fixed":true},"minute":{"value":0,"fixed":true}},"end":{"month":{"value":2,"fixed":false},"day":{"value":1,"fixed":true},"hour":{"value":0,"fixed":true},"minute":{"value":0,"fixed":true}}},"recurrence":{"type":"m","day":15,"hour":17,"minute":45}}}}',
            });
        });
    });
});
