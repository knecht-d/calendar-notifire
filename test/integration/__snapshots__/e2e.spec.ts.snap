// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`End to end test trigger settings should read the config even if user is not an admin 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "chat1",
      "trigger1:
    Jeden Monat am 15. um 17:45
    NÃ¤chte Erinnerung am 15.8.2020 um 17:45 zeigt Termine von 1.9.2020 00:00 bis 1.10.2020 00:00",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
